Add tutorial about running Pods with sidecar containers (#46825)
kubernetes/website: content/en/docs/concepts/workloads/pods/sidecar-containers.md

On Linux, resource allocations for Pod level control groups (cgroups) are based on the effective Pod
request and limit, the same as the scheduler.

> ### Adopting sidecar containers

Reviewer: This section might be more suitable for a reference page rather than a section of the concepts page.

Author: I would appreciate a pointer on which page is better. I didn't think it deserve the whole task page under https://kubernetes.io/docs/tasks/. I am not sure where under reference would be a good place.

This section is relevant for people adopting sidecar containers for their workloads.

> The sidecar containers feature is in beta state starting Kubernetes version 1.29 and is enabled by default.

Reviewer: Better use the shortcode for feature gates.
This sentence becomes invalid when the gate is removed.
By the way, "sidecar containers" is not the name of the gate.

Some clusters may have this feature disabled or have software installed that is incompatible with the feature.
Here are the considerations and troubleshooting steps that one can take while adopting sidecar containers for their workload.

**1. Ensure the feature gate is enabled.**

As a very first step, make sure that both - API server and Nodes were upgraded to the version 1.29+.

> The feature will break on clusters where Nodes are running earlier versions where it is not enabled.

Reviewer: Is it okay to run a 1.28 version Node with the feature (still Alpha) enabled?

Author: added an alert below

The feature gate enablement should also be validated for both - API server AND Nodes.

> Once of the ways to check the feature gate enablement is running a command like this:

Reviewer: Once -> One

Author: ```suggestion
One of the ways to check the feature gate enablement is running a command like this:
```

- For API Server 
  `kubectl get --raw /metrics | grep kubernetes_feature_enabled | grep SidecarContainers`
- For the individual node: 

>   `kubectl get --raw /api/v1/nodes/gke-cluster-device-p-almost-real-tpu--122d9b42-tnb5/proxy/metrics | grep kubernetes_feature_enabled | grep SidecarContainers`

Reviewer: ```suggestion
  `kubectl get --raw /api/v1/nodes/<node-name>/proxy/metrics | grep kubernetes_feature_enabled | grep SidecarContainers`
```

If you see something like: `kubernetes_feature_enabled{name="SidecarContainers",stage="BETA"} 1`,
it means that the feature is enabled.

**2. Check for 3rd party tooling and mutating webhooks**

> If trying the feature you experience issues, it may be an indication that one of the

Reviewer: ```suggestion
If you experience issues when validating the feature, it may be an indication that one of the
```

3rd party tools or mutating webhooks are broken.

> The Sidecar feature adds a new field that is define in Kubernetes 1.28 API.

Reviewer: ```suggestion
The SidecarContainers feature adds a new field that is define in Kubernetes 1.28 API.
```

> Some tools and mutating webhooks might have been build with the earlier version of Kubernetes API.

Reviewer: ```suggestion
Some tools or mutating webhooks might have been built with an earlier version of Kubernetes API.
```

If tools pass the unknown fields as-is or mutating webhooks are using the recommended `Patch` method to mutate Pods, it will not be a problem.

> However there are tooling that will strip out unknown fields and must be recompiled with the fresh version of Kubernetes API.

Reviewer: ```suggestion
However there are tools that will strip out unknown fields and must be recompiled with the v1.28+ version of Kubernetes API.
```

> The easy way to validate if this is a root cause of a problem is to use `kiubectl describe pod` command with your Pod.

Reviewer: ```suggestion
The way to check this is to use the `kubectl describe pod` command with your Pod.
```

> If any tooling stripped out the new field (`restartPolicy:Always`), you will not find see it in the command output.

Reviewer: ```suggestion
If any tools stripped out the new field (`restartPolicy:Always`), you will not find see it in the command output.
```

Reviewer 2: Even better: avoid “new” here. We like [timeless documentation](https://developers.google.com/style/timeless-documentation).

> If you hit an issue like this, please also advice authors of tooling or web hooks to switch to `Patch` method of modifying objects.

Reviewer: ```suggestion
If you hit an issue like this, please advise the author of the tools or the webhooks to switch to `Patch` method of modifying objects.
```

**3. Automatic injection of sidecars**

If you are using software that injects sidecars automatically,
there are a few possible strategies you may follow to
ensure that native sidecar container can be used.

> All of the strategies are generally a ways you may choose to decide whether

Reviewer: ```suggestion
All of the strategies are generally options you may choose to decide whether
```

* Read about [creating a Pod that has an init container](/docs/tasks/configure-pod-container/configure-pod-initialization/#create-a-pod-that-has-an-init-container).
* Learn about the [types of probes](/docs/concepts/workloads/pods/pod-lifecycle/#types-of-probe): liveness, readiness, startup probe.
* Learn about [pod overhead](/docs/concepts/scheduling-eviction/pod-overhead/).

> * Learn how to [Adopt Sidecar Containers](/docs/tutorials/configuration/pod-sidecar-containers/)

Reviewer 2: I'd list this as the first item in the list, not the last/

Author: makes sense. moved
